src/main/scala/is/hail/distributedmatrix/BlockMatrixIsDistributedMatrix.scala

```diff
@@ -0,0 +1,249 @@
+    require(l.numRows() == r.numRows())
+    require(l.numCols() == r.numCols())
+    require(l.rowsPerBlock == r.rowsPerBlock)
+    require(l.colsPerBlock == r.colsPerBlock)
+    val blocks: RDD[((Int, Int), Matrix)] = l.blocks.join(r.blocks).mapValues { case (m1, m2) =>
+      val size = m1.numRows * m1.numCols
+      val result = new Array[Double](size)
+      var j = 0
+      while (j < m1.numCols) {
+        var i = 0
+        while (i < m1.numRows) {
+          result(i + j*m1.numRows) = op(m1(i, j), m2(i, j))
+          i += 1
+        }
+        j += 1
+      }
+      new DenseMatrix(m1.numRows, m1.numCols, result)
+    }
+    new BlockMatrix(blocks, l.rowsPerBlock, l.colsPerBlock, l.numRows(), l.numCols())
+  }
+
+  def pointwiseAdd(l: M, r: M): M = l.add(r)
+  def pointwiseSubtract(l: M, r: M): M = l.subtract(r)
+  def pointwiseMultiply(l: M, r: M): M = map2(_ * _)(l, r)
+  def pointwiseDivide(l: M, r: M): M = map2(_ / _)(l, r)
+
+  def map(op: Double => Double)(m: M): M = {
+    val blocks: RDD[((Int, Int), Matrix)] = m.blocks.mapValues { case m =>
+      val size = m.numRows * m.numCols
+      val result = new Array[Double](size)
+      var j = 0
+      while (j < m.numCols) {
+        var i = 0
+        while (i < m.numRows) {
+          result(i + j*m.numRows) = op(m(i, j))
+          i += 1
+        }
+        j += 1
+      }
+      new DenseMatrix(m.numRows, m.numCols, result)
+    }
+    new BlockMatrix(blocks, m.rowsPerBlock, m.colsPerBlock, m.numRows(), m.numCols())
+  }
+  def scalarAdd(m: M, i: Double): M = map(_ + i)(m)
+  def scalarSubtract(m: M, i: Double): M = map(_ - i)(m)
+  def scalarSubtract(i: Double, m: M): M = map(i - _)(m)
+  def scalarMultiply(m: M, i: Double): M = map(_ * i)(m)
+  def scalarDivide(m: M, i: Double): M = map(_ / i)(m)
+  def scalarDivide(i: Double, m: M): M = map(i / _)(m)
+
+  private def mapWithRowIndex(op: (Double, Int) => Double)(x: M): M = {
+    val nRows: Long = x.numRows
+    val nCols: Long = x.numCols
+    val rowsPerBlock: Int = x.rowsPerBlock
+    val colsPerBlock: Int = x.colsPerBlock
+    val rowBlocks: Int = ((nRows - 1) / rowsPerBlock).toInt + 1
+    val colBlocks: Int = ((nCols - 1) / colsPerBlock).toInt + 1
+    val rowsRemainder: Int = (nRows % rowsPerBlock).toInt
+    val colsRemainder: Int = (nCols % colsPerBlock).toInt
+    val blocks: RDD[((Int, Int), Matrix)] = x.blocks.mapValuesWithKey { case ((blockRow, blockCol), m) =>
+      new DenseMatrix(m.numRows, m.numCols, m.toArray.zipWithIndex.map { case (e, j) =>
+        val rowsInThisBlock: Int = (if (blockRow + 1 == rowBlocks && rowsRemainder != 0) rowsRemainder else rowsPerBlock)
+        val colsInThisBlock: Int = (if (blockCol + 1 == colBlocks && colsRemainder != 0) colsRemainder else colsPerBlock)
+        if (blockRow.toLong * rowsInThisBlock + j % rowsInThisBlock < nRows &&
+          blockCol.toLong * colsInThisBlock + j / rowsInThisBlock < nCols)
+          op(e, blockRow * rowsInThisBlock + j % rowsInThisBlock)
```

Comment:
`blockRow * rowInThisBlock` seems wrong.  Also, why isn't this just `blockRow * rowsPerBlock + j`?  Shouldn't `j` always be `< rowsInThisBlock`?